File: BusinessLogic/Sites/clsFFNet.vb
Imports System
Imports System.text
Imports System.net
Imports System.IO
Imports System.Xml
Imports HtmlAgilityPack

Class FFNet
    Inherits clsFanfic

    Private Browser As clsWeb

#Region "Retrieval Functions"

    Public Overrides _
    Function GrabTitle( _
                        ByVal htmldoc As String _
                      ) As String

        Dim value As String
        Dim doc As HtmlDocument
        Dim ch As Integer = 0
        Dim title As String

        doc = CleanHTML(htmldoc)

        value = doc.DocumentNode.SelectSingleNode("//title").InnerText
        value = Replace(value, "| FanFiction", "")

        ch = InStr(value, "Chapter")

        If ch > 0 Then
            title = Trim(Mid(value, 1, ch - 1))
            value = Replace(value, title, "")
            ch = LastPos(value, ",")
            value = title & "<br><br>" & Mid(value, 1, ch - 1)
            'value = title

        End If

        doc = Nothing

        Return value

    End Function

    Public Overrides _
    Function GrabSeries(ByVal htmlDoc As String) As String

        Dim value As String
        Dim doc As HtmlDocument

        doc = CleanHTML(htmlDoc)

        value = doc.DocumentNode.SelectSingleNode("//title").InnerText
        value = Replace(value, "| FanFiction", "")

        If value <> "" Then
            value = Trim(Mid(value, LastPos(value, ",") + 1))
        End If

        doc = Nothing

        Return value

    End Function

    Public Overrides _
    Function GrabDate( _
                       ByVal htmlDoc As String, _
                       ByVal title As String _
                     ) As String

        Dim ret As String = ""

        Dim author As String = GrabAuthor(htmlDoc)
        Dim rss_link As String = author
        Dim fic As clsFanfic.Story = Nothing

Retry:

        If IsNothing(datasetRSS) Then
            MyBase.GrabFeed(rss_link)
        End If

        fic = GetStoryInfoByID(GetStoryID(StoryURL))

        If LCase(fic.Author) <> LCase(author) Then
            datasetRSS = Nothing
            GoTo Retry
        End If

        Select Case title
            Case "Published: "
                ret = fic.PublishDate
            Case "Updated: "
                If CDate(fic.PublishDate) > CDate(fic.UpdateDate) Then
                    ret = fic.PublishDate
                Else
                    ret = fic.UpdateDate
                End If
        End Select

        Return ret

    End Function

    Public Overrides _
    Function GrabAuthor(ByVal htmldoc As String) As String

        Dim ret As String
        Dim temp As HtmlNodeCollection
        Dim doc As HtmlDocument

        doc = CleanHTML(htmldoc)

        temp = FindLinksByHref(doc.DocumentNode, "/u/")

        ret = temp(0).InnerText

        temp = Nothing
        doc = Nothing

        Return ret

    End Function

    Public Overrides _
    Function GrabBody(ByVal htmldoc As String) As String

        Dim ret As String = ""

        Dim doc As HtmlDocument
        Dim temp As HtmlNodeCollection

        doc = CleanHTML(htmldoc)

        temp = doc.DocumentNode.SelectNodes("//div[@id='storytext']")

        htmldoc = temp(0).InnerHtml

        doc = Nothing
        temp = Nothing

        ret = htmldoc

        Return ret

    End Function

    Public Overrides _
    Function GetChapters( _
                          ByVal htmlDoc As String _
                        ) As String()

        Dim data() As String

        data = GetOptionValues(htmlDoc, "Chapter Navigation")

        If IsNothing(data) Then
            ReDim data(0)
            data(0) = "Chapter 1"
        End If

        Me.Chapters = data

        Return data

    End Function

#End Region

    Public Overrides _
    Function GrabData(ByVal url As String) As String

        Dim htmldoc As String

        htmldoc = MyBase.GrabData(url)

        StripTag(htmldoc, "class", "menu", partialM.Yes)
        StripTag(htmldoc, "clsas", "dropdown", partialM.Yes)

        StripTag(htmldoc, "button")
        StripTag(htmldoc, "link")
        StripTag(htmldoc, "img")
        StripTag(htmldoc, "script")
        StripTag(htmldoc, "style")
        StripTag(htmldoc, "meta")

        Return htmldoc

    End Function

    Protected Overrides Function GrabFeedData(ByRef rss As String) As System.Xml.XmlDocument

        Dim txtatom As String
        Dim txtresult As String
        Dim xmldoc As XmlDocument

        rss = Replace(rss, " ", "")
        rss = Replace(rss, "-", "")
        rss = Replace(rss, "'", "")

        If InStr(rss, Me.HostName) = 0 Then
            rss = Replace(rss, ".", "")
            rss = "http://www.fanfiction.net/~" & rss
        End If

        If (InStr(rss, "atom") = 0) Then

            txtresult = Browser.DownloadPage(rss)

            Try

                txtresult = Mid(txtresult, InStr(txtresult, "id: "))
                txtresult = Mid(txtresult, 1, InStr(txtresult, "<") - 1)
                If InStr(txtresult, ",") > 0 Then
                    txtresult = Mid(txtresult, 1, InStr(txtresult, ",") - 1)
                End If

                txtatom = "/atom/u/"
                txtatom += Replace(txtresult, "id: ", "")

                If txtatom = "" Then rss = ""

                Dim url As URL
                url = ExtractUrl(rss)

                If InStr(txtatom, url.Scheme) = 0 Then
                    txtatom = url.Scheme & "://" & url.Host & txtatom
                End If

                rss = txtatom

            Catch
                rss = ""
            End Try

        End If

        If rss = "" Then
            xmldoc = Nothing
        Else
            xmldoc = DownloadXML(rss)
            xmldoc = CleanXML(xmldoc)
            xmldoc = CleanFeed(xmldoc)
        End If

        If Not IsNothing(xmldoc) Then

        End If

        Return xmldoc

    End Function

    Public Overrides _
    Function InitialDownload(ByVal link As String) As String

        Dim host As String
        Dim htmlDoc As String

        Dim URL As URL

        URL = ExtractUrl(link)

        host = URL.Scheme & "://" & URL.Host & "/s/"
        link = Replace(link, host, "")
        host = host & Mid(link, 1, InStr(link, "/") - 1)
        link = host & "/1/"

        htmlDoc = GrabData(link)

        InitialDownload = htmlDoc

    End Function

    Public Overrides _
    Function ProcessChapters( _
                              ByVal link As String, _
                              ByVal index As Integer _
                            ) As String

        Dim host As String
        Dim temp As String
        Dim params() As String

        Dim url As URL

        url = ExtractUrl(link)

        host = url.Scheme & "://" & url.Host & "/s/"

        temp = Replace(link, host, "")
        params = Split(temp, "/")

        link = host & params(0) & "/"

        Dim htmldoc As String

        htmldoc = GrabData(link & (index + 1) & "/")

        ProcessChapters = htmldoc

    End Function

    Public Overrides Function GrabStoryInfo(ByVal idx As Integer) As clsFanfic.Story

        Dim dsRSS As DataSet = MyBase.datasetRSS

        Dim fic As New clsFanfic.Story

        Dim txtSummary() As String
        Dim txtResult As String
        Dim category As String

        Dim index As Integer

        Dim rss_url As String

        ' Story Name
        fic.Title = dsRSS.Tables("entry"). _
                        Rows(idx).Item("title")



        ' Story Author
        fic.Author = dsRSS.Tables("author"). _
                          Rows(idx).Item(0)

        rss_url = fic.Author

        fic.AuthorURL = GetAuthorURL(rss_url)

        ' Story Location
        fic.StoryURL = dsRSS.Tables("link"). _
                          Rows(idx).Item(1)

        'Process Summary
        txtResult = dsRSS.Tables("summary"). _
                    Rows(idx).Item(1) & vbCrLf

        fic.ID = Me.GetStoryID(fic.StoryURL)

        txtResult = Replace(txtResult, "<hr>", "<br>")
        txtResult = Replace(txtResult, "<hr size=1>", "<br>")

        txtResult = Replace(txtResult, ", Words", "<br>Words")
        txtResult = Replace(txtResult, ", Reviews", "<br>Reviews")

        txtSummary = Split(txtResult, "<br>")

        'Category
        category = Mid(txtSummary(1), Len("Category: "))
        category = Mid(category, (InStr(category, ">") + 1))
        category = Replace(category, "</a>", "")

        fic.Category = category

        If InStr(txtSummary(5), "Words") = 0 Then
            index = 5
        Else
            index = 4
        End If


        'Chapter Count
        fic.ChapterCount = txtSummary(index)

        fic.PublishDate = dsRSS.Tables("entry"). _
                          Rows(idx).Item("published")

        'Last Updated
        fic.UpdateDate = dsRSS.Tables("entry"). _
                         Rows(idx).Item("id")
        fic.UpdateDate = Split(fic.UpdateDate, ",")(1)
        fic.UpdateDate = Split(fic.UpdateDate, ":")(0)
        fic.UpdateDate = CDate(fic.UpdateDate).ToShortDateString

        'Published
        fic.PublishDate = dsRSS.Tables("entry"). _
                          Rows(idx).Item("published")
        fic.PublishDate = CDate(fic.PublishDate).ToShortDateString

        'Summary

        fic.Summary = txtSummary(UBound(txtSummary))

        Return fic

    End Function


    Public Overrides Function GetStoryID(ByVal link As String) As String

        Dim ret As String
        Dim hl As URL
        Dim parms() As String

        hl = ExtractUrl(link)

        parms = Split(hl.URI, "/")

        ret = parms(2)

        Return ret

    End Function

    Public Overrides Function GetStoryURL(ByVal id As String) As String

        Dim link As String

        link = "http://www.fanfiction.net/s/" & id & "/1/"

        Return link

    End Function

    Public Overrides Function GetAuthorURL(ByVal link As String) As String

        Dim ret As String

        If InStr(link, Me.HostName) = 0 Then
            link = Replace(link, ".", "")
            link = "http://www.fanfiction.net/~" & link
        End If

        ret = Replace(link, "atom/", "")

        Return ret

    End Function

#Region "Properties"

    Public Overrides ReadOnly Property ErrorMessage() As String
        Get
            Return "story not found"
        End Get
    End Property

    Public Overrides ReadOnly Property HostName() As String
        Get
            Return "fanfiction.net"
        End Get
    End Property

    Public Overrides ReadOnly Property Name() As String
        Get
            Return "FFNet"
        End Get
    End Property

#End Region

    Public Sub New()

        Browser = New clsWeb

    End Sub
End Class

File: Downloader/HtmlScraper/BusinessLogic/Sites/clsFFNet.cs
using System.Xml;
using HtmlAgilityPack;
using HtmlScraper.BusinessLogic.Sites.Base;
using HtmlScraper.Utility;
using HtmlScraper.Utility.Browser.HTML;
using HtmlScraper.Utility.Browser.XML;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace HtmlScraper.BusinessLogic.Sites
{

    class FFNet : clsFanfic
    {

        #region Retrieval Functions

        public override string GrabTitle(string htmldoc)


        {

            string value;
            HtmlAgilityPack.HtmlDocument doc;
            int ch = 0;
            string title;

            doc = modHTML.CleanHTML(ref htmldoc);

            value = doc.DocumentNode.SelectSingleNode("//title").InnerText;
            value = Strings.Replace(value, "| FanFiction", "");

            ch = Strings.InStr(value, "Chapter");

            if (ch > 0)
            {
                title = Strings.Trim(Strings.Mid(value, 1, ch - 1));
                value = Strings.Replace(value, title, "");
                ch = modUtility.LastPos(value, ",");
                value = title + "<br><br>" + Strings.Mid(value, 1, ch - 1);
                // value = title

            }

            doc = null;

            return value;

        }

        public override string GrabSeries(string htmlDoc)
        {

            string value;
            HtmlAgilityPack.HtmlDocument doc;

            doc = modHTML.CleanHTML(ref htmlDoc);

            value = doc.DocumentNode.SelectSingleNode("//title").InnerText;
            value = Strings.Replace(value, "| FanFiction", "");

            if (!string.IsNullOrEmpty(value))
            {
                value = Strings.Trim(Strings.Mid(value, modUtility.LastPos(value, ",") + 1));
            }

            doc = null;

            return value;

        }

        public override string GrabDate(string htmlDoc, string title)



        {

            string ret = "";

            string author = GrabAuthor(htmlDoc);
            string rss_link = author;
            Story fic = default;

        Retry:
            ;


            if (datasetRSS == null)
            {
                GrabFeed(ref rss_link);
            }

            fic = GetStoryInfoByID(GetStoryID(StoryURL));

            if ((Strings.LCase(fic.Author) ?? "") != (Strings.LCase(author) ?? ""))
            {
                datasetRSS = null;
                goto Retry;
            }

            switch (title ?? "")
            {
                case "Published: ":
                    {
                        ret = fic.PublishDate;
                        break;
                    }
                case "Updated: ":
                    {
                        if (Conversions.ToDate(fic.PublishDate) > Conversions.ToDate(fic.UpdateDate))
                        {
                            ret = fic.PublishDate;
                        }
                        else
                        {
                            ret = fic.UpdateDate;
                        }

                        break;
                    }
            }

            return ret;

        }

        public override string GrabAuthor(string htmldoc)
        {

            string ret;
            HtmlNodeCollection temp;
            HtmlAgilityPack.HtmlDocument doc;

            doc = modHTML.CleanHTML(ref htmldoc);

            temp = modHTML.FindLinksByHref(doc.DocumentNode, "/u/");

            ret = temp[0].InnerText;

            temp = null;
            doc = null;

            return ret;

        }

        public override string GrabBody(string htmldoc)
        {

            string ret = "";

            HtmlAgilityPack.HtmlDocument doc;
            HtmlNodeCollection temp;

            doc = modHTML.CleanHTML(ref htmldoc);

            temp = doc.DocumentNode.SelectNodes("//div[@id='storytext']");

            htmldoc = temp[0].InnerHtml;

            doc = null;
            temp = null;

            ret = htmldoc;

            return ret;

        }

        public override string[] GetChapters(string htmlDoc)


        {

            string[] data;

            data = modHTML.GetOptionValues(htmlDoc, "Chapter Navigation");

            if (data == null)
            {
                data = new string[1];
                data[0] = "Chapter 1";
            }

            Chapters = data;

            return data;

        }

        #endregion

        public override string GrabData(string url)
        {

            string htmldoc;

            htmldoc = base.GrabData(url);

            modStripTags.StripTag(ref htmldoc, "class", "menu", modStripTags.partialM.Yes);
            modStripTags.StripTag(ref htmldoc, "clsas", "dropdown", modStripTags.partialM.Yes);

            modStripTags.StripTag(ref htmldoc, "button");
            modStripTags.StripTag(ref htmldoc, "link");
            modStripTags.StripTag(ref htmldoc, "img");
            modStripTags.StripTag(ref htmldoc, "script");
            modStripTags.StripTag(ref htmldoc, "style");
            modStripTags.StripTag(ref htmldoc, "meta");

            return htmldoc;

        }

        protected override XmlDocument GrabFeedData(ref string rss)
        {

            string txtatom;
            string txtresult;
            XmlDocument xmldoc;

            rss = Strings.Replace(rss, " ", "");
            rss = Strings.Replace(rss, "-", "");
            rss = Strings.Replace(rss, "'", "");

            if (Strings.InStr(rss, HostName) == 0)
            {
                rss = Strings.Replace(rss, ".", "");
                rss = "http://www.fanfiction.net/~" + rss;
            }

            if (Strings.InStr(rss, "atom") == 0)
            {

                txtresult = Program.Browser.DownloadPage(rss);

                try
                {

                    txtresult = Strings.Mid(txtresult, Strings.InStr(txtresult, "id: "));
                    txtresult = Strings.Mid(txtresult, 1, Strings.InStr(txtresult, "<") - 1);
                    if (Strings.InStr(txtresult, ",") > 0)
                    {
                        txtresult = Strings.Mid(txtresult, 1, Strings.InStr(txtresult, ",") - 1);
                    }

                    txtatom = "/atom/u/";
                    txtatom += Strings.Replace(txtresult, "id: ", "");

                    if (string.IsNullOrEmpty(txtatom))
                        rss = "";

                    URL url;
                    url = URLHelper.ExtractUrl(rss);

                    if (Strings.InStr(txtatom, url.Scheme) == 0)
                    {
                        txtatom = url.Scheme + "://" + url.Host + txtatom;
                    }

                    rss = txtatom;
                }

                catch
                {
                    rss = "";
                }

            }

            if (string.IsNullOrEmpty(rss))
            {
                xmldoc = null;
            }
            else
            {
                xmldoc = modXML.DownloadXML(rss);
                xmldoc = modXML.CleanXML(xmldoc);
                xmldoc = modRSS.CleanFeed(xmldoc);
            }

            if (!(xmldoc == null))
            {

            }

            return xmldoc;

        }

        public override string InitialDownload(string link)
        {
            string InitialDownloadRet = default;

            string host;
            string htmlDoc;

            URL URL;

            URL = URLHelper.ExtractUrl(link);

            host = URL.Scheme + "://" + URL.Host + "/s/";
            link = Strings.Replace(link, host, "");
            host = host + Strings.Mid(link, 1, Strings.InStr(link, "/") - 1);
            link = host + "/1/";

            htmlDoc = GrabData(link);

            InitialDownloadRet = htmlDoc;
            return InitialDownloadRet;

        }

        public override string ProcessChapters(string link, int index)



        {
            string ProcessChaptersRet = default;

            string host;
            string temp;
            string[] @params;

            URL url;

            url = URLHelper.ExtractUrl(link);

            host = url.Scheme + "://" + url.Host + "/s/";

            temp = Strings.Replace(link, host, "");
            @params = Strings.Split(temp, "/");

            link = host + @params[0] + "/";

            string htmldoc;

            htmldoc = GrabData(link + (index + 1) + "/");

            ProcessChaptersRet = htmldoc;
            return ProcessChaptersRet;

        }

        public override Story GrabStoryInfo(int idx)
        {

            var dsRSS = datasetRSS;

            var fic = new Story();

            string[] txtSummary;
            string txtResult;
            string category;

            int index;

            string rss_url;

            // Story Name
            fic.Title = Conversions.ToString(dsRSS.Tables["entry"].Rows[idx]["title"]);



            // Story Author
            fic.Author = Conversions.ToString(dsRSS.Tables["author"].Rows[idx][0]);

            rss_url = fic.Author;

            fic.AuthorURL = GetAuthorURL(rss_url);

            // Story Location
            fic.StoryURL = Conversions.ToString(dsRSS.Tables["link"].Rows[idx][1]);

            // Process Summary
            txtResult = Conversions.ToString(Operators.ConcatenateObject(dsRSS.Tables["summary"].Rows[idx][1], Constants.vbCrLf));

            fic.ID = GetStoryID(fic.StoryURL);

            txtResult = Strings.Replace(txtResult, "<hr>", "<br>");
            txtResult = Strings.Replace(txtResult, "<hr size=1>", "<br>");

            txtResult = Strings.Replace(txtResult, ", Words", "<br>Words");
            txtResult = Strings.Replace(txtResult, ", Reviews", "<br>Reviews");

            txtSummary = Strings.Split(txtResult, "<br>");

            // Category
            category = Strings.Mid(txtSummary[1], Strings.Len("Category: "));
            category = Strings.Mid(category, Strings.InStr(category, ">") + 1);
            category = Strings.Replace(category, "</a>", "");

            fic.Category = category;

            if (Strings.InStr(txtSummary[5], "Words") == 0)
            {
                index = 5;
            }
            else
            {
                index = 4;
            }


            // Chapter Count
            fic.ChapterCount = txtSummary[index];

            fic.PublishDate = Conversions.ToString(dsRSS.Tables["entry"].Rows[idx]["published"]);

            // Last Updated
            fic.UpdateDate = Conversions.ToString(dsRSS.Tables["entry"].Rows[idx]["id"]);
            fic.UpdateDate = Strings.Split(fic.UpdateDate, ",")[1];
            fic.UpdateDate = Strings.Split(fic.UpdateDate, ":")[0];
            fic.UpdateDate = Conversions.ToDate(fic.UpdateDate).ToShortDateString();

            // Published
            fic.PublishDate = Conversions.ToString(dsRSS.Tables["entry"].Rows[idx]["published"]);
            fic.PublishDate = Conversions.ToDate(fic.PublishDate).ToShortDateString();

            // Summary

            fic.Summary = txtSummary[Information.UBound(txtSummary)];

            return fic;

        }


        public override string GetStoryID(string link)
        {

            string ret;
            URL hl;
            string[] parms;

            hl = URLHelper.ExtractUrl(link);

            parms = Strings.Split(hl.URI, "/");

            ret = parms[2];

            return ret;

        }

        public override string GetStoryURL(string id)
        {

            string link;

            link = "http://www.fanfiction.net/s/" + id + "/1/";

            return link;

        }

        public override string GetAuthorURL(string link)
        {

            string ret;

            if (Strings.InStr(link, HostName) == 0)
            {
                link = Strings.Replace(link, ".", "");
                link = "http://www.fanfiction.net/~" + link;
            }

            ret = Strings.Replace(link, "atom/", "");

            return ret;

        }

        #region Properties

        public override string ErrorMessage
        {
            get
            {
                return "story not found";
            }
        }

        public override string HostName
        {
            get
            {
                return "fanfiction.net";
            }
        }

        public override string Name
        {
            get
            {
                return "FFNet";
            }
        }

        #endregion

    }
}